File: LogisticRegression/java/LogisticRegression/src/LogisticRegression.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import org.ejml.data.DenseMatrix64F;
import org.ejml.simple.SimpleMatrix;


public class LogisticRegression {
	public static void main(String[] args){
		SimpleMatrix trainDataSimpleMatrix = loadData("train.txt");
		SimpleMatrix testDataSimpleMatrix = loadData("test.txt");
		SimpleMatrix weight = new SimpleMatrix(1,21);
		SimpleMatrix updateWeight = weight;
		double rate = 0.001;
		for (int i=0;i<2000;i++){
			updateWeight = updateWeight.minus(gradientDescent(weight,trainDataSimpleMatrix).scale(rate));
			weight = updateWeight;
			
		}
		System.out.println(updateWeight);
	}
	public static double sign(double n){
		if (n<=0)
			return -1;
		else
			return 1;
	}
	public static SimpleMatrix gradientDescent(SimpleMatrix w,SimpleMatrix data){
		int rowNum = data.numRows();
		int colNum = data.numCols();
		SimpleMatrix xdata = data.extractMatrix(0, rowNum, 0, colNum-1);
		SimpleMatrix ydata = data.extractVector(false, colNum-1);
		SimpleMatrix weight = new SimpleMatrix(1,colNum-1);
		
		for (int i=0;i<rowNum;i++){
			double coef = 1/(1+Math.exp(-1*ydata.get(i)*w.dot(xdata.extractVector(true, i))));
			weight = weight.plus(xdata.extractVector(true, i).scale(ydata.get(i)*(-1)*coef));
		}
		
		return weight.scale((double) 1/rowNum);
		
	}
	public static SimpleMatrix loadData(String fileName){
		List<String> dataList = new ArrayList<String>();
		BufferedReader br = null;
		try {
				br = new BufferedReader(new FileReader(fileName));
				String line = null;
				while((line = br.readLine())!=null){
					dataList.add(line.trim());
				}
				br.close();
		} catch (IOException e){
			e.printStackTrace();
		}
		int rowNum = dataList.size();
		int colNum = dataList.get(0).split("\\s+").length+1;
		double[] dataMatrix = new double[rowNum*colNum];
		for (int i=0;i<rowNum;i++){
			dataMatrix[i*colNum] = 1;
			String[] newLine = dataList.get(i).split("\\s+");
			//System.out.println(dataList.get(i));
			for (int j=0;j<newLine.length;j++){
				dataMatrix[i*colNum+j+1] = Double.valueOf(newLine[j]);
			}
		}
		DenseMatrix64F dataDenseMatrix = new DenseMatrix64F(rowNum,colNum,true,dataMatrix);
		SimpleMatrix dataSimpleMatrix = SimpleMatrix.wrap(dataDenseMatrix);
		return dataSimpleMatrix;
	}
}

File: LogisticRegression/py/LogisticRegression.py
import numpy as np
import math
def load_data(train):
	try:
		matrix=[]
		f = open(train,'r')
		for line in f.readlines():
			x = list(map(float,line.split()))
			x.insert(0,1.0)
			matrix.append(x)
	finally:
		if f:
			f.close()
	return matrix

def load_test(test):
	try:
		matrix=[]
		f = open(test,'r')
		for line in f.readlines():
			x = list(map(float,line.split()))
			x.insert(0,1.0)
			matrix.append(x)
	finally:
		if f:
			f.close()
	return matrix

def logistic(x):
	return 1/(1+math.exp(-x))

def sign (n):
	if n > 0:
		return 1
	else :
		return -1

def predict(row,weight):
	colNum = len(row)-1
	x = row[:colNum]
	y = row[colNum]
	if sign(np.dot(x,weight))!=y:
		return 1
	else:
		return 0	

def gradientDescent(data,w):
	update = w
	gradientupdate = w
	rowNum = len(data)
	colNum = len(data[0])
	for i in range(rowNum):
		x = data[i][:colNum-1]
		y = data[i][colNum-1]
		coef = logistic(-y*np.dot(update,np.transpose(x)))
		gradientupdate = np.add(gradientupdate,np.dot(-y*coef,x))
	return np.dot(1/rowNum,gradientupdate)

train = load_data('train.txt')
test = load_data('test.txt')
weight = []
for i in range(len(train[0])-1):
	weight.append(0)
learningRate = 0.001
for i in range(2000):
	update = gradientDescent(train,weight)
	w = np.subtract(weight,np.dot(update,learningRate))
	weight = w
	print('weight:',weight)
errCount = 0
for i in test:
	errCount = errCount + predict(i,weight)
print("out sample error rate:",errCount/len(test))


